OCPBUGS-7089, OCPBUGS-7113: move cluster menu to masthead to fix usability issues

rhamilto opened the pull request:
Note I removed the icon from the menu items as Joy's latest design omits them and it improves the readability of the menu toggle, especially at mobile.

Known differences in the masthead:
* the masthead background color override (black) is removed, so the masthead matches PatternFly (near black)
* the masthead is 70px tall instead of 76px due to the difference in PatternFly's `<Masthead>` vs.`<PageHeader>` 

Without cluster menu and nearly all masthead toolbar icons:

https://user-images.githubusercontent.com/895728/215136439-915145e0-139f-4255-ac3d-0ddc7033dfd8.mov

With cluster menu and relevant masthead toolbar icons:

https://user-images.githubusercontent.com/895728/216684107-a2cc062d-dfc9-4e10-ac1e-7dd4dc8365e8.mov

spadgett commented:
Thanks @rhamilto 

> upon initial load, the masthead briefly renders the mobile view before switching to desktop when at desktop resolutions

Let's open a PF bug for this.

Should we still collapse the icons in the kebab at mobile? Looks like we might have room to leave some or all of the icons.

You've probably tested this already, but we should check how that notification drawer and quick starts work as expected.

rhamilto commented:
> Thanks @rhamilto
> 
> > upon initial load, the masthead briefly renders the mobile view before switching to desktop when at desktop resolutions
> Let's open a PF bug for this.

I think this is something specific to console.  Needs further investigation.

> Should we still collapse the icons in the kebab at mobile? Looks like we might have room to leave some or all of the icons.

I don't think we'll have room once the cluster picker is there as there is an optional `SystemStatusButton` and the notification bell if one or more alerts are present (in other words, there may be two icons beside the kebab, and those take up the available space).
 
> You've probably tested this already, but we should check how that notification drawer and quick starts work as expected.

They were both ok in my testing.  Am gonna get a shared environment running so y'all can help test (clusterbot failed).

rhamilto commented:
> > Thanks @rhamilto
> > > upon initial load, the masthead briefly renders the mobile view before switching to desktop when at desktop resolutions
> > > Let's open a PF bug for this.
> 
> I think this is something specific to console. Needs further investigation.

This was a configuration issue with `<Masthead>`.  See https://github.com/openshift/console/pull/12480/commits/872e14e7576a7a34c9e34076da86c032eb7b3fa8#diff-405eb65f788e5a0d149321eec7b359f9e6c22a54c1fd7f1973b2bc6d508f599cL71

There is a flicker of the kebab on initial load now though.  Haven't figured out a workaround.  Maybe it's ok?

https://user-images.githubusercontent.com/895728/214948180-ab6e3996-0b58-4b4b-9408-7e1883382132.mov

rhamilto commented:
/retest

rhamilto commented:
> > > upon initial load, the masthead briefly renders the mobile view before switching to desktop when at desktop resolutions
> > > Let's open a PF bug for this.
> 
> I think this is something specific to console. Needs further investigation.

I identified and added a workaround.

> > Should we still collapse the icons in the kebab at mobile? Looks like we might have room to leave some or all of the icons.
> 
> I don't think we'll have room once the cluster picker is there as there is an optional `SystemStatusButton` and the notification bell if one or more alerts are present (in other words, there may be two icons beside the kebab, and those take up the available space).

I updated the video above to include the optional icons.

jcaianirh reviewed (changes requested):
@rhamilto https://github.com/openshift/console/pull/12366 has merged into master

rhamilto commented:
/retest

rhamilto commented:
/jira refresh

rhamilto commented:
/retest

rhamilto commented:
/retest

rhamilto commented:
/retest

sg00dwin commented:
@rhamilto did the design specify the selector, within the masthead, straddle the sidenav bar and main content areas? At <1200 it could left edge align with the main content, like the project selector.
I feel like Clayton would fixate on that as well. 😅   

<img width="590" alt="Screenshot 2023-02-08 at 12 37 22 PM" src="https://user-images.githubusercontent.com/1874151/217608533-822c537a-b242-488b-ac30-7b84dc53c672.png">

rhamilto commented:
> @rhamilto did the design specify the selector, within the masthead, straddle the sidenav bar and main content areas? At <1200 it could left edge align with the main content, like the project selector. I feel like Clayton would fixate on that as well. 😅
> 
> <img alt="Screenshot 2023-02-08 at 12 37 22 PM" width="590" src="https://user-images.githubusercontent.com/1874151/217608533-822c537a-b242-488b-ac30-7b84dc53c672.png">

There isn't a design per se.  This is the default layout behavior of PatternFly's `<Masthead>`.  We certainly could add CSS to do what you're suggesting though.

sg00dwin commented:
/retest-required

rhamilto commented:
/jira refresh

rhamilto commented:
/retest

rhamilto commented:
/retest

rhamilto commented:
@jhadvig or @TheRealJon, mind reviewing?

TheRealJon commented on frontend/public/components/app.jsx:
```diff
@@ -58,7 +58,8 @@ import '@patternfly/quickstarts/dist/quickstarts.min.css';
 import '@patternfly/patternfly/patternfly-charts-theme-dark.css';
 
-const breakpointMD = 1200;
+const PF_BREAKPOINT_MD = 768;
+const PF_BREAKPOINT_XL = 1200;
 const NOTIFICATION_DRAWER_BREAKPOINT = 1800;
```
Are these values, or comparable values, available from PF of PF React packages?

rhamilto replied:
Good thought!  It doesn't appear so:
* https://github.com/search?q=repo%3Apatternfly%2Fpatternfly-react%201200&type=code
* https://github.com/search?q=repo%3Apatternfly%2Fpatternfly-react+768&type=code

jcaianirh replied:
@nicolethoen do you know if the page breakpoints that pf advises are available for export from the pf packages?

nicolethoen replied:
I believe you can import CSS variable names and values via @patternfly/react-tokens.
For example
`import globalBreakpointLg from '@patternfly/react-tokens/dist/esm/global_breakpoint_lg';`
and in that case, `globalBreakpointLg` contains three piece of information
```
name: "--pf-global--breakpoint--lg"
value: "992px"
var: "var(--pf-global--breakpoint--lg)"
```

rhamilto replied:
Thanks, @nicolethoen.  Unfortunately I need only a number (e.g., `1200`) and not a number with a unit (e.g., `1200px`), so I'll have to stick with what I've got.

sg00dwin commented on frontend/packages/console-app/src/components/nav/ClusterMenuToggle.tsx:
```diff
@@ -52,7 +52,7 @@ const ClusterMenuToggle = (props: {
       onClick={() => onToggle(!isOpen)}
       isExpanded={isOpen}
       variant="plainText"
```
```suggestion
```
If we're trying to be consistent with console.redhat.com, then `Menu` toggle would have a standard display with `hover` and `active` blue border bottom. This can be achieved by removing this and modifying some styles in the _masthead.scss

rhamilto replied:
Yeah.  I chose not to include that since we have a regular dropdown and not a full-page one.  Seemed like it might be weird?

rhamilto replied:
@sg00dwin and I discussed this, and are opting to leave it as is as removing requires more CSS overrides.

sg00dwin commented on frontend/public/components/_masthead.scss:
```diff
@@ -1,3 +1,56 @@
+  }
+
+  // PatternFly does not fully support <MenuToggle isFullHeight> styling in <Masthead>
+  &::before {
+    border: 1px solid var(--pf-global--palette--black-800) !important;
+    border-bottom: 0 !important;
+    border-top: 0 !important;
+    bottom: 0;
+    content: "";
+    left: 0;
+    position: absolute;
+    right: 0;
+    top: 0;
+  }
```
```suggestion
  &::before {
    bottom: 0;
    content: "";
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
  }
  
  &.pf-c-menu-toggle {
    --pf-c-menu-toggle--after--BorderBottomColor: transparent;
  }
```

sg00dwin replied:
`--pf-global--palette--black-800` doesn't contrast on the dark theme header. The default menu toggle borders seem to work better.

rhamilto replied:
Updated to add a better border color for dark mode.

sg00dwin commented on frontend/public/components/_masthead.scss:
```diff
@@ -1,3 +1,56 @@
+}
+
+.co-cluster-selector {
+  display: inline-grid !important;
```
```suggestion
  background-color: transparent !important;
  display: inline-grid !important;
```

rhamilto replied:
Not needed since we're not removing `variant="plain"`.

sg00dwin commented on frontend/public/components/_masthead.scss:
```diff
@@ -1,3 +1,56 @@
+.co-cluster-menu {
+  max-width: 80vw;
+
+  @media (min-width: $screen-sm) {
```
```suggestion
  @media (min-width: $pf-global--breakpoint--md) {
```

rhamilto replied:
Updated

sg00dwin commented on frontend/public/components/_masthead.scss:
```diff
@@ -1,3 +1,56 @@
+    max-width: 300px !important;
+  }
+
+  @media (min-width: $screen-lg) {
```
```suggestion
  @media (min-width: $pf-global--breakpoint--xl) {
```
`$screen-lg` === `$pf-global--breakpoint--xl` => 1200

rhamilto replied:
Leave as is.

rhamilto commented:
/hold

found a bug

sg00dwin commented on frontend/public/components/_masthead.scss:
```diff
@@ -1,3 +1,56 @@
+    max-width: 180px !important;
+  }
+
+  @media (min-width: $screen-xs) {
```
```suggestion
  @media (min-width: 480px) {
```

rhamilto replied:
Leave as is.

rhamilto commented:
@TheRealJon, @sg00dwin.  I fixed a couple of issues.  PTAnotherLook.

TheRealJon reviewed (approved):
/lgtm

rhamilto commented:
/hold cancel

rhamilto commented:
/retest

rhamilto commented:
/retest
